### PR TITLE
Update timestamps using diff alignment algorithm

### PR DESCRIPTION
**Pull request by murezzda**
**Is your Pull Request request related to another [issue](https://github.com/bbc/react-transcript-editor/issues) in this repository ?**      
Related to #30

**Describe what the PR does**    
To restore timestamps after editing of the transcript, the new state is compared to the original state of the transcript. [diff-match-patch](https://github.com/google/diff-match-patch) is used to find all differences between the original and new transcript, and the following rules are applied:

- If a new word matches the original word, move the timestamps to the new word
- If a word is removed, ignore the original timestamp.
- If a word is added, set the timestamp to the previous original word (in this case, two consecutive words will have the same timestamp).
- If multiple words are replaced by a single word, set the timestamp of that word to the combined timestamp of all replaced words.
- If multiple words are replaced by multiple words, set all timestamps of the new words to the combined timestamp of all replaced words.

**State whether the PR is ready for review or whether it needs extra work**    
The following points need to be clarified
- The point of updating the timestamps is currently a timer (like the save document timer). The timestamp update can take some time, especially on larger files, thus it might be better to trigger this by hand.
- Saving the "original" transcript might need some adjustments.

### CODE REVIEW

**Comment by pietrop** [comment]
Thanks @murezzda , very interesting approach!

There's a module, we just open sourced [`stt-align-node`](https://github.com/bbc/stt-align-node), latest changes are on `diffs-report` branch in [`diffs-report/lib/index.js`](https://github.com/bbc/stt-align-node/blob/diffs-report/lib/index.js) which does a similar diffing, with a different diffing library. (Originally developed by @chrisbaume in BBC R&D in python) - apologies for not open sourcing it sooner.

See more details in this [comment](https://github.com/bbc/react-transcript-editor/issues/30#issuecomment-449045769) in this issue https://github.com/bbc/react-transcript-editor/issues/30 on how I've tried to use it with the `TranscriptEditor`.

In this case that module is able to do an interpolation of the new inserted words, and transpose the time-codes for the matching once. As you can see explained in the [README](https://github.com/bbc/stt-align-node/tree/diffs-report)

I think you are onto something, and there might be a way to combine elements of these two things for an optimal approach.

**Comment by Laurian** [comment]
Ideally we could devise later (as in: accept this PR and refactor later) a mechanism to swap implementations of various timestamp correcting things, like a plugin style or something. Such that we could have: this, stt-align-node, remote API call (Gentle, Aeneas, Speechmatics aligner)

**Comment by murezzda** [comment]
Thanks for the input! I think for the next step I will check out `stt-align-node`. If it seems feasible, I will integrate it in my code.

**Comment by murezzda** [comment]
I've added now the `stt-align-node` code and use it as the current method to update the timestamps. The `diff-match-patch` method is still in the PR, but currently not used.

What I've experienced so far with `stt-align-node` is that it works generally better for insertion and mutliple words substitiutions, but at the beginning of a longer pause words can have timestamps far to long into the pause instead at the beginning of the pause. This might be improved by combining `stt-align-node` with `diff-match-patch`, which I want to try next.

**Comment by pietrop** [comment]
Thanks @murezzda !

See if this can help with the testing, on the Ted website, there is the[ "human accurate" transcript for this ted Talk](https://www.ted.com/talks/kate_darling_why_we_have_an_emotional_connection_to_robots/transcript?language=en).

Here for convenience, with line spaces, and time-codes removed
```
There was a day, about 10 years ago, when I asked a friend to hold a baby dinosaur robot upside down. It was this toy called a Pleo that I had ordered, and I was really excited about it because I've always loved robots. And this one has really cool technical features. It had motors and touch sensors and it had an infrared camera. And one of the things it had was a tilt sensor, so it knew what direction it was facing. And when you held it upside down, it would start to cry. And I thought this was super cool, so I was showing it off to my friend, and I said, "Oh, hold it up by the tail. See what it does." So we're watching the theatrics of this robot struggle and cry out. And after a few seconds, it starts to bother me a little, and I said, "OK, that's enough now. Let's put him back down." And then I pet the robot to make it stop crying.
And that was kind of a weird experience for me. For one thing, I wasn't the most maternal person at the time. Although since then I've become a mother, nine months ago, and I've learned that babies also squirm when you hold them upside down.
(Laughter)
But my response to this robot was also interesting because I knew exactly how this machine worked, and yet I still felt compelled to be kind to it. And that observation sparked a curiosity that I've spent the past decade pursuing. Why did I comfort this robot? And one of the things I discovered was that my treatment of this machine was more than just an awkward moment in my living room, that in a world where we're increasingly integrating robots into our lives, an instinct like that might actually have consequences, because the first thing that I discovered is that it's not just me.
In 2007, the Washington Post reported that the United States military was testing this robot that defused land mines. And the way it worked was it was shaped like a stick insect and it would walk around a minefield on its legs, and every time it stepped on a mine, one of the legs would blow up, and it would continue on the other legs to blow up more mines. And the colonel who was in charge of this testing exercise ends up calling it off, because, he says, it's too inhumane to watch this damaged robot drag itself along the minefield. Now, what would cause a hardened military officer and someone like myself to have this response to robots?
Well, of course, we're primed by science fiction and pop culture to really want to personify these things, but it goes a little bit deeper than that. It turns out that we're biologically hardwired to project intent and life onto any movement in our physical space that seems autonomous to us. So people will treat all sorts of robots like they're alive. These bomb-disposal units get names. They get medals of honor. They've had funerals for them with gun salutes. And research shows that we do this even with very simple household robots, like the Roomba vacuum cleaner.
(Laughter)
It's just a disc that roams around your floor to clean it, but just the fact it's moving around on its own will cause people to name the Roomba and feel bad for the Roomba when it gets stuck under the couch.
(Laughter)
And we can design robots specifically to evoke this response, using eyes and faces or movements that people automatically, subconsciously associate with states of mind. And there's an entire body of research called human-robot interaction that really shows how well this works. So for example, researchers at Stanford University found out that it makes people really uncomfortable when you ask them to touch a robot's private parts.
(Laughter)
So from this, but from many other studies, we know, we know that people respond to the cues given to them by these lifelike machines, even if they know that they're not real.
Now, we're headed towards a world where robots are everywhere. Robotic technology is moving out from behind factory walls. It's entering workplaces, households. And as these machines that can sense and make autonomous decisions and learn enter into these shared spaces, I think that maybe the best analogy we have for this is our relationship with animals. Thousands of years ago, we started to domesticate animals, and we trained them for work and weaponry and companionship. And throughout history, we've treated some animals like tools or like products, and other animals, we've treated with kindness and we've given a place in society as our companions. I think it's plausible we might start to integrate robots in similar ways.
And sure, animals are alive. Robots are not. And I can tell you, from working with roboticists, that we're pretty far away from developing robots that can feel anything. But we feel for them, and that matters, because if we're trying to integrate robots into these shared spaces, we need to understand that people will treat them differently than other devices, and that in some cases, for example, the case of a soldier who becomes emotionally attached to the robot that they work with, that can be anything from inefficient to dangerous. But in other cases, it can actually be useful to foster this emotional connection to robots. We're already seeing some great use cases, for example, robots working with autistic children to engage them in ways that we haven't seen previously, or robots working with teachers to engage kids in learning with new results. And it's not just for kids. Early studies show that robots can help doctors and patients in health care settings.
This is the PARO baby seal robot. It's used in nursing homes and with dementia patients. It's been around for a while. And I remember, years ago, being at a party and telling someone about this robot, and her response was, "Oh my gosh. That's horrible. I can't believe we're giving people robots instead of human care." And this is a really common response, and I think it's absolutely correct, because that would be terrible. But in this case, it's not what this robot replaces. What this robot replaces is animal therapy in contexts where we can't use real animals but we can use robots, because people will consistently treat them more like an animal than a device.
Acknowledging this emotional connection to robots can also help us anticipate challenges as these devices move into more intimate areas of people's lives. For example, is it OK if your child's teddy bear robot records private conversations? Is it OK if your sex robot has compelling in-app purchases?
(Laughter)
Because robots plus capitalism equals questions around consumer protection and privacy.
And those aren't the only reasons that our behavior around these machines could matter. A few years after that first initial experience I had with this baby dinosaur robot, I did a workshop with my friend Hannes Gassert. And we took five of these baby dinosaur robots and we gave them to five teams of people. And we had them name them and play with them and interact with them for about an hour. And then we unveiled a hammer and a hatchet and we told them to torture and kill the robots.
(Laughter)
And this turned out to be a little more dramatic than we expected it to be, because none of the participants would even so much as strike these baby dinosaur robots, so we had to improvise a little, and at some point, we said, "OK, you can save your team's robot if you destroy another team's robot."
(Laughter)
And even that didn't work. They couldn't do it. So finally, we said, "We're going to destroy all of the robots unless someone takes a hatchet to one of them." And this guy stood up, and he took the hatchet, and the whole room winced as he brought the hatchet down on the robot's neck, and there was this half-joking, half-serious moment of silence in the room for this fallen robot.
(Laughter)
So that was a really interesting experience. Now, it wasn't a controlled study, obviously, but it did lead to some later research that I did at MIT with Palash Nandy and Cynthia Breazeal, where we had people come into the lab and smash these HEXBUGs that move around in a really lifelike way, like insects. So instead of choosing something cute that people are drawn to, we chose something more basic, and what we found was that high-empathy people would hesitate more to hit the HEXBUGS.
Now this is just a little study, but it's part of a larger body of research that is starting to indicate that there may be a connection between people's tendencies for empathy and their behavior around robots. But my question for the coming era of human-robot interaction is not: "Do we empathize with robots?" It's: "Can robots change people's empathy?" Is there reason to, for example, prevent your child from kicking a robotic dog, not just out of respect for property, but because the child might be more likely to kick a real dog?
And again, it's not just kids. This is the violent video games question, but it's on a completely new level because of this visceral physicality that we respond more intensely to than to images on a screen. When we behave violently towards robots, specifically robots that are designed to mimic life, is that a healthy outlet for violent behavior or is that training our cruelty muscles? We don't know ... But the answer to this question has the potential to impact human behavior, it has the potential to impact social norms, it has the potential to inspire rules around what we can and can't do with certain robots, similar to our animal cruelty laws. Because even if robots can't feel, our behavior towards them might matter for us. And regardless of whether we end up changing our rules, robots might be able to help us come to a new understanding of ourselves.
Most of what I've learned over the past 10 years has not been about technology at all. It's been about human psychology and empathy and how we relate to others. Because when a child is kind to a Roomba, when a soldier tries to save a robot on the battlefield, or when a group of people refuses to harm a robotic baby dinosaur, those robots aren't just motors and gears and algorithms. They're reflections of our own humanity.
Thank you.
(Applause) 
```

I've also noticed that the colour highlight in `TimedTextEditor` while playing is not a great measure to check the accuracy of the alignment, as it might drag behind a little bit, best bet, is to double click on a word, and see if corresponding sound comes out. 

alignement

Other than that, seems pretty accurate!

**Comment by murezzda** [comment]
Update for this branch:
- I've tried to merge the two attempts, but as I discovered they are not as easily compatible as I though. For the moment, I wont persuade this method anymore.
- Updating the timestamps causes the focus to be lost. This is not that big of a problem at long intervals between timestamp updating, but on shorter intervals it is no longer practical to edit the text.

**Comment by pietrop** [comment]
sounds good, re not merging the two diffing approaches.

As for point 2, it could either be done on a "final save" (eg just before sending to the server) or more granularly at paragraph level, eg once user moves onto second paragraph, and see if this doesn't loose the focus. Or the focus could be restored after the update.

I also think there might be tweaks to address this issue https://github.com/bbc/react-transcript-editor/issues/150 that will effect this PR.

Thanks @murezzda will have a closer look as soon as I get a chance.

**Comment by murezzda** [comment]
Selection is now preserved when the timestamp is updated, which improves the usability significantly. The update of the selection state is done by mapping the old selection state on the new state. Performance wise the impact of moving the selection state seems neglible.

**Comment by murezzda** [comment]
Regarding the overall performance of the update, it still introduces a lag if the file is long (such as the Kate Darling TED Talk).

I took some measurements and it seems the alignment of the timestamps is not the main factor, the conversion and updating of the state also takes at least as much time as the alignment: 

Conversion of state to Raw: 33.01611328125ms
STT alignment: 47.970947265625ms
Conversion of state from Raw: 58.0439453125ms
Update editor state: 49.7158203125ms

As previously mentioned, updating the selection takes about 42.7 ms in total and thus has about the same impact on the performance as all other functions used here.

**Comment by pietrop** [comment]
Thanks for this @murezzda
Very interesting the measurements, I wonder if there' s a way to narrow down the update to only the paragraph where the text has changed? and whether that would make any difference?

**Comment by murezzda** [comment]
Hi @pietrop

The problem is that the conversion to and from raw anyway have to happen, thus it would only be possible to save time in the STT alignment section. I've also thought about checking if something changed in the transcript and only align if that is the case, but it looks to me that obtaining the information where and if something has changed is more or less as expensive as just aligning the text itself.

Whats your opinion, should we try to improve the performance or could we also handle this by placing the call to the realignment at some different point in the application? At the moment, the realignment happens whenever the current state of the transkript is saved.

Regarding the measurements, I've used the console.time('test') and console.timeEnd('test') methods. Do you know if these are reliable for react?

**Comment by pietrop** [comment]
Hi @murezzda,
I think placing it on save could be a good choice. 

Altho perhaps instead of the automatic save( that happens when people stop typing), we could place it after the manual save btn?

And add a way so that when a parent component exports the content, a realignment is done if text has been changed since the last one - if that makes sense.

Re measurements, other might be able to advise? (@Laurian  @jamesdools @barneyboo )

In this PR https://github.com/bbc/react-transcript-editor/pull/160 where we looked at the unnecessary re-renders I made [some notes](https://github.com/bbc/react-transcript-editor/blob/fix-159-performance-problem/docs/notes/2019-05-16-prevent-unnecessary-re-renders-in-react.md) and one of the things I found was the [React DevTools tools profiler](https://github.com/bbc/react-transcript-editor/blob/fix-159-performance-problem/docs/notes/2019-05-16-prevent-unnecessary-re-renders-in-react.md#react-devtools-tools-profiler) - see [in react docs](https://reactjs.org/blog/2018/09/10/introducing-the-react-profiler.html) but I am not sure if/how to use that to measure that type of performance? but might be worth looking into it.

**Comment by murezzda** [comment]
Hi @pietrop 

I've now tried to put the update as a timer into the  `if (this.state.editorState.getCurrentContent() !== editorState.getCurrentContent())` code of the `onChange` function. This imporves the user-experience, as alignments only happen if something has changed in the text.

Whats a bit of a problem with this is that it can be that unaliged text can be automatically saved, but as the alignment anyways happens on the "original" transcript, this should not be that big of a problem. Especially if the alignment is also triggered by the manual save button, this gives the user a way to trigger the aligment explicitly.

Let me know what you think of this. If this would be a viable solution, I would clean up the code and give it up for review.

**Comment by pietrop** [comment]
@murezzda this sounds great!
Feel free to clean it up and let me know when is ready for review, I can make some time next week to go through it 😄

**Comment by murezzda** [comment]
@pietrop cleanup done, ready for review.

**Comment by pietrop** [comment]
Easiest fix for being more certain when retrieving data from the outer component that there are word level timings for each word is to make the restoring time-codes part of the export process.

Motivation as mentioned above is so that when saving to server there’s more certainty on whether there are time-codes for each word.

Possible solution is to add a step in [`getEditorContent`](https://github.com/bbc/react-transcript-editor/pull/144/files#diff-3035aeb31edb298cacc61b804909d4ebR171) to `updateTimestampsForEditorState`.

```diff
  getEditorContent(exportFormat) {
    const format = exportFormat || 'draftjs';
++ this.updateTimestampsForEditorState();
    return exportAdapter(convertToRaw(this.state.editorState.getCurrentContent()), format);
  }
```

If done this way, it could be part of this PR.

---

Just for background context, the current way of getting the editor's content from a parent component  can be seen in demo app where the `exportTranscript` function  is uses `refs` and `getEditorContent` function.
``` js
const { data, ext } = this.transcriptEditorRef.current.getEditorContent(
      this.state.exportFormat
    );
```

File: package.json [hunk 1]
```diff
@@ -46,8 +46,12 @@
     "@fortawesome/free-solid-svg-icons": "^5.6.3",
     "@fortawesome/react-fontawesome": "^0.1.3",
     "babel-polyfill": "^6.26.0",
+    "diff-match-patch": "^1.0.4",
```
[the remark this answers is not in the document]
Comment (murezzda) [reply package.json]: Yes, this is from the previous attempt to fix the timecodes via the diff-match-patch lib. Thanks for finding this.

File: packages/components/timed-text-editor/UpdateTimestamps/index.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      start: parseFloat(entityMap[entityIdx].data.start),
+      end: parseFloat(entityMap[entityIdx].data.end),
+      word: entityMap[entityIdx].data.text,
+    });
+  }
+
+  const result = alignWords( entities, currentText);
+
+  const newEntities = result.map((entry) => {
+    return createEntity(entry.start, entry.end, 0.0, entry.word, -1);
```
[the remark this answers is not in the document]
Comment (murezzda) [reply packages/components/timed-text-editor/UpdateTimestamps/index.js]: Yes, this would be a fix. At the moment the index is not set in any other way.

**Comment by pietrop** [comment]
https://github.com/murezzda/react-transcript-editor/compare/update-timestamps-diff...bbc:murezzda-update-timestamps-diff

**Comment by pietrop** [comment]
I've run a test with this interview From PBS Frontline Documentary [The Facebook Dilemma, Soleio Cuervo, Former Facebook Product Designer](https://www.pbs.org/wgbh/frontline/interview/soleio-cuervo)

### steps to reproduce
1. Accurate text of transcription and video from [youtube](https://www.youtube.com/watch?v=OIAUfZBd_7w) via [`electron-video-downloader`](https://github.com/pietrop/electron-video-downloader)

- [`The Facebook Dilemma - Soleio Cuervo-OIAUfZBd_7w.en.txt`](https://github.com/bbc/react-transcript-editor/files/3396846/The.Facebook.Dilemma.-.Soleio.Cuervo-OIAUfZBd_7w.en.txt)

2. Run the transcript via [autoEdit](http://www.autoedit.io) using BBC Kaldi (can use any other STT provider tho) and exported as json
- [`The_Facebook Dilemma - Soleio Cuervo-OIAUfZBd_7w.mp4_12_07_2019_15-27-47-autoEdit2-from-Octo-Kaldi.json.txt`](https://github.com/bbc/react-transcript-editor/files/3396849/The_Facebook.Dilemma.-.Soleio.Cuervo-OIAUfZBd_7w.mp4_12_07_2019_15-27-47-autoEdit2-from-Octo-Kaldi.json.txt) (remove the `.txt` to turn it back into a json)

3. Imported in `react-transcript-editor` - local version for this  branch [`murezzda-update-timestamps-diff`](https://github.com/bbc/react-transcript-editor/tree/murezzda-update-timestamps-diff) based on this PR #144  - [`http://localhost:6006`](http://localhost:6006/?path=/story/demo--default) storybook demo page. (choose `autoEdit2` under import transcript options.)

4. Imported the media downloaded form youtube.

###  results
Screen Shot 2019-07-16 at 12 36 12

The alignment was very good, all the way to the end. 🎉 

### Caveats
With this algorithm the alignment is only as good as the accuracy of the STT.
The accuracy of STT recognition is effected by 
- quality of the STT engine 
- quality of the recording, eg microphone used, background noise, distance of speaker from mic etc..
- (presumably) enunciation of words and accents (?)

To try this first hand, used `pocketsphinx` STT (free, open source and works offline) option in [autoEdit](http://www.autoedit.io) to transcribe the same recording.
- [`The_Facebook Dilemma - Soleio Cuervo-OIAUfZBd_7w.mp4_12_07_2019_15-27-47-autoEdit2-from-pocketsphinx.json.txt](https://github.com/bbc/react-transcript-editor/files/3396884/The_Facebook.Dilemma.-.Soleio.Cuervo-OIAUfZBd_7w.mp4_12_07_2019_15-27-47-autoEdit2-from-pocketsphinx.json.txt`) (remove `.txt` to convert back to `json`)

It starts good, and then It drifts out of sync, incrementally. It is out of sync by a paragraph or more by the end of the interview.

For this edge case, where the STT or the recognition might be a bit off, we might not have enough insights on why it's going out of sync, is it just the diffing that doesn't work? is there something that can be added to the algo to handle that etc..?
Or another thought is could we recognise smaller chunks at a time? altho not sure how we'd segment for that?

File: packages/components/timed-text-editor/UpdateTimestamps/index.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  for (var blockIdx in content.blocks) {
+    const block = content.blocks[blockIdx];
+    const blockArray = block.text.match(/\S+/g) || [];
+    text = text.concat(blockArray);
+  }
+
+  return (text);
+};
+
+const createEntity = (start, end, confidence, word, wordIdx) => {
```
Comment (jamesdools) [inline packages/components/timed-text-editor/UpdateTimestamps/index.js]: wordIdx -> wordIndex, or index

File: packages/components/timed-text-editor/UpdateTimestamps/index.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    confidence: confidence,
+    word: word.toLowerCase().replace(/[.?!]/g, ''),
+    punct: word,
+    index: wordIdx,
+  });
+};
+
+const createContentFromEntityList = (currentContent, newEntities) => {
+  // Update entites to block structure.
+  var updatedBlockArray = [];
```
Comment (jamesdools) [inline packages/components/timed-text-editor/UpdateTimestamps/index.js]: not using var, as per Pietro's comment 👍

**Comment by pietrop** [comment]
Awesome! thanks @jamesdools I'll have a closer look and add those changes as a separate PR to https://github.com/bbc/react-transcript-editor/pull/175 branch tomorrow

File: packages/components/timed-text-editor/UpdateTimestamps/index.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import generateEntitiesRanges from '../../../stt-adapters/generate-entities-ranges/index.js';
+import { createEntityMap } from '../../../stt-adapters/index.js';
+import alignWords from './stt-align-node.js';
+
+const convertContentToText = (content) => {
+  var text = [];
+
+  for (var blockIdx in content.blocks) {
```
[the remark this answers is not in the document]
Comment (pietrop) [reply packages/components/timed-text-editor/UpdateTimestamps/index.js]: going for `blockIndex`

File: packages/components/timed-text-editor/index.js [hunk 6]
```diff
@@ -116,9 +146,45 @@ class TimedTextEditor extends React.Component {
...
+
+    // Set the updated selection state on the new editor state
+    const newEditorStateSelected = EditorState.forceSelection(newEditorState, selection);
+    this.setState({ editorState: newEditorStateSelected });
+  }
+
   loadData() {
     if (this.props.transcriptData !== null) {
       const blocks = sttJsonAdapter(this.props.transcriptData, this.props.sttJsonType);
+      this.setState({ originalState: convertToRaw(convertFromRaw(blocks)) });
```
[the remark this answers is not in the document]
Comment (pietrop) [reply packages/components/timed-text-editor/index.js]: we need to check with @murezzda what was the thinking around this, I am not quiet sure. 

this feels like it was needed as a deep copy? but can't remember why?
````js 
   this.setState({ originalState: convertToRaw(convertFromRaw(blocks)) });
```


and `originalState` could be declared in constructor as well?

File: packages/components/timed-text-editor/UpdateTimestamps/stt-align-node.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+  // transcriptWordsStripped = [re.sub('[^a-z]', '', word.lower()) for word in transcriptWords]
+  const transcriptWordsStripped = transcriptWords.map((word) => {
+    return normaliseWord(word);
+  });
+  // # create empty list to receive data
+  // transcriptData = [{} for _ in range(len(transcriptWords))]
+  const transcriptData = [];
+  // empty objects as place holder
+  transcriptWords.forEach(() => {
```
[the remark this answers is not in the document]
Comment (pietrop) [reply packages/components/timed-text-editor/UpdateTimestamps/stt-align-node.js]: Yeah, I think it's coz originally this algo is mirroring the python code form @chrisbaume  https://github.com/bbc/stt-align - https://github.com/bbc/stt-align-node

Ideally we could tweak it, refactor it there, and then bring it in via npm.
So leaving this for now, before they start to diverge too much.

**Comment by pietrop** [comment]
addressed @jamesdools comments in PR https://github.com/bbc/react-transcript-editor/pull/144  at commit https://github.com/bbc/react-transcript-editor/pull/144#pullrequestreview-273740995 so closing this PR

**Pull request closed without merging**